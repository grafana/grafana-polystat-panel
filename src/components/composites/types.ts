import { SelectableValue } from '@grafana/data';

export interface CompositeMember {}

export const DisplayModes: SelectableValue[] = [
  { value: 'all', label: 'Show All' },
  { value: 'triggered', label: 'Show Triggered' },
];

export interface CompositeMetric {
  seriesMatch: string;
  compositeMatch?: CompositeItemType[];
  alias?: string;
  ID?: string;
  order: number;
}

export interface CompositeItemType {
  name: string;
  label: string;
  order: number;
  templatedName: string;
  isTemplated: boolean;
  displayMode: string;
  enabled: boolean;
  showName: boolean;
  showValue: boolean;
  showComposite: boolean;
  showMembers: boolean;
  metrics: CompositeMetric[];
  clickThrough: string | '';
  clickThroughSanitize: boolean;
  clickThroughOpenNewTab: boolean;
}

export interface CompositeItemTracker {
  composite: CompositeItemType;
  order: number;
  ID: string;
}

export interface CompositeItemProps {
  composite: CompositeItemType;
  ID: string;
  enabled: boolean;
  setter: any;
  remover: any;
  moveUp: any;
  moveDown: any;
  createDuplicate: any;
}

export interface CompositeMetricItemProps {
  metric: CompositeMetric;
  index: number;
  disabled: boolean;
  removeMetric: any;
  updateMetric: any;
  updateMetricAlias: any;
}
