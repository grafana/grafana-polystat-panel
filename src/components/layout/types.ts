
export interface LayoutPoint {
  x: number,
  y: number,
}
