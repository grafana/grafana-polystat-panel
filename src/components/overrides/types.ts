import { PolystatThreshold } from 'components/thresholds/types';

export interface OverrideItemProps {
  override: OverrideItemType;
  ID: string;
  enabled: boolean;
  setter: any;
  remover: any;
  moveUp: any;
  moveDown: any;
  createDuplicate: any;
}

export interface OverrideItemType {
  label: string;
  metricName: string;
  alias: string;
  thresholds: PolystatThreshold[];
  colors: string[];
  unitFormat: string;
  decimals: string;
  scaledDecimals: number | null;
  enabled: boolean;
  operatorName: string; // mean/min/max etc
  prefix: string;
  suffix: string;
  clickThrough: string | '';
  clickThroughSanitize: boolean;
  clickThroughOpenNewTab: boolean;
  order: number;
}

export interface OverrideItemTracker {
  override: OverrideItemType;
  order: number;
  ID: string;
}
