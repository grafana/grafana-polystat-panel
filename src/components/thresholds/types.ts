import { SelectableValue } from '@grafana/data';

export interface PolystatThreshold {
  color: string;
  state: number;
  value: number;
}

export interface ThresholdItemTracker {
  threshold: PolystatThreshold;
  order: number;
  ID: string;
}

export const ThresholdStates: SelectableValue[] = [
  { value: 0, label: 'ok' },
  { value: 1, label: 'warning' },
  { value: 2, label: 'critical' },
  { value: 3, label: 'custom' },
];
