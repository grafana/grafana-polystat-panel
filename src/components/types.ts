import { SelectableValue } from '@grafana/data';
import { CompositeItemType } from 'components/composites/types';
import { OverrideItemType } from './overrides/types';
import { PolystatThreshold } from './thresholds/types';

export interface PolystatOptions {
  autoSizeColumns: boolean;
  autoSizeRows: boolean;
  autoSizePolygons: boolean;
  ellipseCharacters: number;
  ellipseEnabled: boolean;
  globalAutoScaleFonts: boolean;
  globalClickthrough: string;
  globalClickthroughNewTabEnabled: boolean;
  globalClickthroughSanitizedEnabled: boolean;
  globalDecimals: number;
  globalDisplayMode: string;
  globalDisplayTextTriggeredEmpty: string;
  globalFillColor: string;
  globalFontSize: number;
  globalGradientsEnabled: boolean;
  globalOperator: string;
  globalPolygonBorderSize: number;
  globalPolygonBorderColor: string;
  globalPolygonSize: number;
  globalRegexPattern: string;
  globalShape: PolygonShapes;
  globalShowValueEnabled: boolean;
  globalTextFontColor: string;
  globalTextFontAutoColor: string;
  globalTextFontAutoColorEnabled: boolean;
  globalThresholdsConfig: PolystatThreshold[];
  globalTooltipsEnabled: boolean;
  globalTooltipsShowTimestampEnabled: boolean;
  globalUnitFormat: string;
  layoutDisplayLimit: number;
  layoutNumColumns: number;
  layoutNumRows: number;
  processedData?: PolystatModel[] | null;
  panelHeight: any;
  panelWidth: any;
  panelId: number;
  radius: number | null;
  renderTime?: Date;
  sortByField: string;
  sortByDirection: number;
  overrideConfig: {
    overrides: OverrideItemType[];
  };
  compositeConfig: {
    animationSpeed: string;
    composites: CompositeItemType[];
    enabled: boolean;
  };
  tooltipPrimarySortDirection: number;
  tooltipPrimarySortByField: string;
  tooltipSecondarySortDirection: number;
  tooltipSecondarySortByField: string;
  tooltipDisplayMode: string;
  tooltipDisplayTextTriggeredEmpty: string;
}

export interface PolystatModel {
  displayMode?: string;
  thresholdLevel?: number; // 0 = ok, 1 = warn, 2 = crit, 3 = unknown (same as sensu)
  value: number;
  valueFormatted: string;
  valueRounded: number;
  stats: any;
  name: string;
  displayName: string; // Used for composites
  timestamp: number;
  prefix: string;
  suffix: string;
  color: string;
  clickThrough: string;
  operatorName: string;
  newTabEnabled: boolean;
  sanitizedURL: string;
  sanitizeURLEnabled: boolean;
  showName: boolean;
  showValue: boolean;
  isComposite: boolean;
  members: PolystatModel[];
  triggerCache?: any; // holds animation frame info
}

export enum PolygonShapes {
  HEXAGON_POINTED_TOP = 'hexagon_pointed_top',
  CIRCLE = 'circle',
  SQUARE = 'square',
}

export const PolygonNamedShapes = [
  { value: 'hexagon_pointed_top', label: 'Hexagon Pointed Top' },
  { value: 'circle', label: 'Circle' },
  { value: 'square', label: 'Square' },
];

export interface PolystatDiameters {
  diameterX: number;
  diameterY: number;
}

export const OperatorOptions: SelectableValue[] = [
  { value: 'mean', label: 'Mean (avg)' },
  { value: 'sum', label: 'Sum' },
  { value: 'min', label: 'Min' },
  { value: 'max', label: 'Max' },
  { value: 'last', label: 'Last' },
  { value: 'lastNotNull', label: 'Last Not Null' },
  { value: 'first', label: 'First' },
  { value: 'firstNotNull', label: 'First Not Null' },
  { value: 'count', label: 'Count' },
  { value: 'allIsNull', label: 'All Is Null (boolean)' },
  { value: 'allIsZero', label: 'All Is Zero (boolean)' },
  { value: 'delta', label: 'Delta' },
  { value: 'diff', label: 'Difference' },
  { value: 'diffperc', label: 'Difference (Percent)' },
  { value: 'last_time', label: 'Time of Last Point' },
  { value: 'logmin', label: 'Log Min' },
  { value: 'name', label: 'Name' },
  { value: 'nonNullCount', label: 'Non Null Count' },
  { value: 'previousDeltaUp', label: 'Previous Delta Up' },
  { value: 'range', label: 'Range' },
  { value: 'step', label: 'Step' },
];

export const SortOptions = [
  { value: 0, label: 'Disabled' },
  { value: 1, label: 'Alphabetical (asc)' },
  { value: 2, label: 'Alphabetical (desc)' },
  { value: 3, label: 'Numerical (asc)' },
  { value: 4, label: 'Numerical (desc)' },
  { value: 5, label: 'Alphabetical (case-insensitive, asc)' },
  { value: 6, label: 'Alphabetical (case-insensitive, desc)' },
];

export const SortOptionOperators = ['disabled', 'asc', 'desc', 'asc', 'desc', 'iasc', 'idesc'];

export const SortByFieldOptions = [
  { value: 'name', label: 'Name' },
  { value: 'thresholdLevel', label: 'Threshold Level' },
  { value: 'value', label: 'Value' },
];

export const DisplayModes = [
  { value: 'all', label: 'Show All' },
  { value: 'triggered', label: 'Show Triggered' },
];

export const AnimationModes = [
  { value: 'all', text: 'Show All' },
  { value: 'triggered', text: 'Show Triggered' },
];
