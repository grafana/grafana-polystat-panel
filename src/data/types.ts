export const CUSTOM_SPLIT_DELIMITER = '#️⃣🔠🆗🆗🔠#️⃣';
