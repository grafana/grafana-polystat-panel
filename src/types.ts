export interface PolystatThreshold {
  color: string;
  state: number;
  value: number;
}

export class MetricOverride {
  label: string;
  metricName: string;
  thresholds: PolystatThreshold[];
  colors: string[];
  unitFormat: string;
  decimals: string;
  scaledDecimals: number;
  enabled: boolean;
  operatorName: string; // avg/min/max etc
  prefix: string;
  suffix: string;
  clickThrough: string;
  newTabEnabled: boolean;
  sanitizeURLEnabled: boolean;
  sanitizedURL: string;
}

export interface PolystatConfigs {
  animationSpeed: number;
  columns: any;
  columnAutoSize: boolean;
  displayLimit: number;
  defaultClickThrough: '';
  defaultClickThroughNewTab: boolean;
  defaultClickThroughSanitize: boolean;
  fontAutoScale: boolean;
  fontSize: number;
  fontType: string;
  globalUnitFormat: string;
  globalDecimals: number;
  globalDisplayMode: string;
  globalOperatorName: string;
  globalDisplayTextTriggeredEmpty: string;
  globalThresholds?: PolystatThreshold[];
  gradientEnabled: boolean;
  hexagonSortByDirection: string;
  hexagonSortByField: string;
  maxMetrics: number;
  polygonBorderSize: number;
  polygonBorderColor: string;
  polygonGlobalFillColor: string;
  radius: any;
  radiusAutoSize: boolean;
  rows: any;
  rowAutoSize: boolean;
  shape: string;
  tooltipDisplayMode: string;
  tooltipDisplayTextTriggeredEmpty: string;
  tooltipFontSize: number;
  tooltipFontType: string;
  tooltipPrimarySortDirection: string;
  tooltipPrimarySortField: string;
  tooltipSecondarySortDirection: string;
  tooltipSecondarySortField: string;
  tooltipTimestampEnabled: boolean;
}
