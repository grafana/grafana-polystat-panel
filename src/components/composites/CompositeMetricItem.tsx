import React, { useState } from 'react';

import { Input, Select, Field, IconButton, HorizontalGroup } from '@grafana/ui';
import { CompositeMetricItemProps } from './types';
import { SelectableValue } from '@grafana/data';

export const CompositeMetricItem: React.FC<CompositeMetricItemProps> = (options) => {
  const [customOptions, setCustomOptions] = useState<Array<SelectableValue<string>>>([]);

  const xoptions: SelectableValue[] = [
    { label: 'Basic option', value: 0 },
    { label: 'Option with description', value: 1, description: 'this is a description' },
    {
      label: 'Option with description and image',
      value: 2,
      description: 'This is a very elaborate description, describing all the wonders in the world.',
      imgUrl: 'https://placekitten.com/40/40',
    },
  ];

  console.log('Composite Metric Item');
  console.log(options.metric);
  console.log(options.metric.seriesMatch);
  async function copySelectedMetricToClipboard(index: number) {
    if (options.metric.seriesMatch?.value) {
      const aValue = options.metric.seriesMatch.value;
      if ('clipboard' in navigator) {
        if (aValue) {
          return await navigator.clipboard.writeText(aValue);
        }
      } else {
        if (aValue) {
          return document.execCommand('copy', true, aValue);
        }
      }
    }
  }

  const updateMetric = (v: SelectableValue) => {
    options.updateMetric(options.index, v);
  };

  return (
    <HorizontalGroup>
      <IconButton
        key="deleteMetric"
        variant="destructive"
        name="trash-alt"
        tooltip="Delete this metric"
        onClick={() => options.removeMetric(options.index)}
      />
      <IconButton
        key="copyMetric"
        variant="secondary"
        name="copy"
        tooltip="Copy Metric/Regex"
        onClick={() => copySelectedMetricToClipboard(options.index)}
      />
      <Field label="Metric/RegEx" disabled={options.disabled}>
        <Select
          width={24}
          key={`cmi-index-${options.index}`}
          allowCustomValue
          options={[...xoptions, ...customOptions]}
          value={options.metric.seriesMatch}
          onCreateOption={(v) => {
            const customValue: SelectableValue = { value: v, label: v, description: 'custom regex' };
            setCustomOptions([...customOptions, customValue]);
            updateMetric(customValue);
          }}
          onChange={(v) => {
            updateMetric(v);
          }}
        />
      </Field>
      <Field label="Alias" disabled={options.disabled}>
        <Input value={options.metric.alias} placeholder="" />
      </Field>
    </HorizontalGroup>
  );
};
