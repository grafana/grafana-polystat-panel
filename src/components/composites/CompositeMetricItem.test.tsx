import React from 'react';
import { render } from '@testing-library/react';

import { CompositeMetricItem } from './CompositeMetricItem';
import { CompositeMetricItemProps, CompositeMetric } from './types';
import { FieldType, toDataFrame } from '@grafana/data';

describe('Test CompositeMetricItem', () => {
  const time = new Date().getTime();
  const frameA = toDataFrame({
    fields: [
      { name: 'time', type: FieldType.time, values: [time, time + 1, time + 2] },
      { name: 'A-series', type: FieldType.number, values: [200, 210, 220] },
    ],
  });

  const aMetric: CompositeMetric = {
    seriesMatch: '.*',
    order: 0,
  }
  const props: CompositeMetricItemProps = {
    metric: aMetric,
    index: 0,
    disabled: false,
    removeMetric: undefined,
    updateMetric: undefined,
    updateMetricAlias: undefined,
    context: { data : [frameA] },
  };
  beforeEach(() => { });

  describe('Metric Hints', () => {
    it('returns set of hint from labels', () => {
      const { container } = render(
          <CompositeMetricItem {...props} />
      );
      console.log(container.innerHTML);
      expect(container.innerHTML).toMatchSnapshot();
    });
  });
});
