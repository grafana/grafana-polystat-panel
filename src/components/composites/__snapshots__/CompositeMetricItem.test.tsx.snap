// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CompositeMetricItem Metric Hints returns set of hint from labels 1`] = `"<fieldset class="css-1qipljx"><div class="css-ve64a7-horizontal-group" style="width: 100%; height: 100%;"><div class="css-cvef6c-layoutChildrenWrapper"><button aria-label="Delete this metric" class="css-yvi4bi" tabindex="0"><div class="css-wf08df-Icon"><svg xmlns="http://www.w3.org/2000/svg" data-testid="trash-alt" viewBox="0 0 24 24"></svg></div></button></div><div class="css-cvef6c-layoutChildrenWrapper"><button aria-label="Copy Metric/Regex" class="css-x1vujn" tabindex="0"><div class="css-wf08df-Icon"><svg xmlns="http://www.w3.org/2000/svg" data-testid="copy" viewBox="0 0 24 24"></svg></div></button></div><div class="css-cvef6c-layoutChildrenWrapper"><div class="css-8e5b3" id="cmi-field-index-0" style="min-width: 175px;"><div class="css-jt4xma-Label"><label><div class="css-xhqy0o">Metric/RegEx</div></label></div><div><div><div class="css-4339yg"><div class="css-xcstkt-input-wrapper" data-testid="input-wrapper"><div class="css-1w5c5dq-input-inputWrapper"><input class="css-1mlczho-input-input" placeholder="" style="padding-right: 12px;" value=".*"><div class="css-7y3u6k-input-suffix"><div class="css-wf08df-Icon"><svg xmlns="http://www.w3.org/2000/svg" data-testid="angle-down" viewBox="0 0 24 24"></svg></div></div></div></div></div></div></div></div></div><div class="css-cvef6c-layoutChildrenWrapper"><div class="css-8e5b3"><div class="css-jt4xma-Label"><label for="cmi-alias-index-0"><div class="css-xhqy0o">Alias</div></label></div><div><div class="css-xcstkt-input-wrapper" data-testid="input-wrapper"><div class="css-1w5c5dq-input-inputWrapper"><input class="css-1mlczho-input-input" id="cmi-alias-index-0" placeholder="" value=""></div></div></div></div></div></div></fieldset>"`;
